include/camp/resource/sycl.hpp

```diff
@@ -44,94 +44,86 @@ namespace resources
+        } else {
+          if (prevContextIter == queueMap_end) {
+            static sycl::context privateContext;
+            syclContext = &privateContext;
           }
         }
-        m_mtx.unlock();
 
-        static int previous = 0;
+        if (prevContextIter == queueMap_end || num < 0) {
+          std::lock_guard<std::mutex> lock(s_mtx);
+
+          if (prevContextIter == queueMap_end) {
+            auto contextIter = queueMap.find(syclContext);
+            if (contextIter == queueMap_end) {
+              static constexpr auto gpuSelector = sycl::gpu_selector_v;
+              static const sycl::property_list propertyList =
+                  sycl::property_list(sycl::property::queue::in_order());
+
+              contextIter = queueMap.emplace(syclContext, { num_queues-1, {
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList)}}).first;
+            }
+            prevContextIter = contextIter;
+          }
 
-        static std::once_flag m_onceFlag;
-        CAMP_ALLOW_UNUSED_LOCAL(m_onceFlag);
-        if (num < 0) {
-          m_mtx.lock();
-          previous = (previous + 1) % 16;
-          m_mtx.unlock();
-          return &queueMap[contextInUse][previous];
+          if (num < 0) {
+            int& previous = prevContextIter->second.first;
+            previous = (previous + 1) % num_queues;
+            return &prevContextIter->second.second[previous];
+          }
         }
 
-        return &queueMap[contextInUse][num % 16];
+        return &prevContextIter->second.second[num % num_queues];
       }
 
       // Private from-queue constructor
       Sycl(sycl::queue& q) : qu(&q) {}
 
     public:
       Sycl(int group = -1)
+        : qu(get_a_queue(nullptr, group))
       {
-        sycl::context temp;
-        qu = get_a_queue(temp, group, false);
       }
 
       Sycl(sycl::context &syclContext, int group = -1)
```

Comment:
Separate question, should the order of these arguments be changed to be more like the other resources that have (group, device)?

Reply:
I added a constructor with this ordering, but didn't remove this one.